lib/types/ipblock.go

```diff
@@ -0,0 +1,201 @@
+	}
+	return block, nil
+}
+
+func ipBlockFromTwoIPs(ip0, ip1 net.IP) *ipBlock {
+	// This code doesn't do any checks on the validity of the arguments, that should be
+	// done before and/or after it is called
+	var block ipBlock
+	block.firstIP = new(big.Int)
+	block.count = new(big.Int)
+	block.ipv6 = ip0.To4() == nil
+	if block.ipv6 {
+		block.firstIP.SetBytes(ip0.To16())
+		block.count.SetBytes(ip1.To16())
+	} else {
+		block.firstIP.SetBytes(ip0.To4())
+		block.count.SetBytes(ip1.To4())
+	}
+	block.count.Sub(block.count, block.firstIP)
+	block.count.Add(block.count, big.NewInt(1))
+
+	return &block
+}
+
+func ipBlockFromCIDR(s string) (*ipBlock, error) {
+	_, pnet, err := net.ParseCIDR(s)
+	if err != nil {
+		return nil, fmt.Errorf("parseCIDR() failed parsing %s: %w", s, err)
+	}
+	ip0 := pnet.IP
+	// TODO: this is just to copy it, it will probably be better to copy the bytes ...
+	ip1 := net.ParseIP(ip0.String())
+	if ip1.To4() == nil {
+		ip1 = ip1.To16()
+	} else {
+		ip1 = ip1.To4()
+	}
+	for i := range ip1 {
+		ip1[i] |= (255 ^ pnet.Mask[i])
+	}
+	block := ipBlockFromTwoIPs(ip0, ip1)
+	// in the case of ipv4 if the network is bigger than 31 the first and last IP are reserved so we
+	// need to reduce the addresses by 2 and increment the first ip
+	if !block.ipv6 && big.NewInt(2).Cmp(block.count) < 0 {
+		block.count.Sub(block.count, big.NewInt(2))
+		block.firstIP.Add(block.firstIP, big.NewInt(1))
+	}
+	return block, nil
+}
+
+func (b ipPoolBlock) getIP(index *big.Int) net.IP {
+	// TODO implement walking ipv6 networks first
+	// that will probably require more math ... including knowing which is the next network and ...
+	// thinking about it - it looks like it's going to be kind of hard or badly defined
+	i := new(big.Int)
+	i.Add(b.firstIP, index)
+	// TODO use big.Int.FillBytes when golang 1.14 is no longer supported
+	return net.IP(i.Bytes())
+}
+
+// NewIPPool returns an IPPool slice from the provided string representation that should be comma
+// separated list of IPs, IP ranges(ip1-ip2) and CIDRs
+func NewIPPool(ranges string) (*IPPool, error) {
+	ss := strings.Split(strings.TrimSpace(ranges), ",")
+	pool := &IPPool{}
+	pool.list = make([]ipPoolBlock, len(ss))
+	pool.count = new(big.Int)
+	for i, bs := range ss {
+		r, err := getIPBlock(bs)
+		if err != nil {
+			return nil, err
+		}
+
+		pool.list[i] = ipPoolBlock{
+			firstIP:    r.firstIP,
+			startIndex: new(big.Int).Set(pool.count), // this is how many there are until now
+		}
+		pool.count.Add(pool.count, r.count)
+	}
+
+	// The list gets reversed here as later it is searched based on when the index we are looking is
+	// bigger than startIndex but it will be true always for the first block which is with
+	// startIndex 0. This can also be fixed by iterating in reverse but this seems better
+	for i := 0; i < len(pool.list)/2; i++ {
+		pool.list[i], pool.list[len(pool.list)/2-i] = pool.list[len(pool.list)/2-i], pool.list[i]
```

Comment:
a reverse should be:  pool.list[i], pool.list[len(pool.list)-i] = pool.list[len(pool.list)-i], pool.list[i]